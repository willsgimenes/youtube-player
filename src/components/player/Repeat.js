import React from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { setLoop } from '../../actions'

import { saveState } from '../../Helpers/Localstorage'

function mapStateToProps ({ player: { loop } }) {
  return {
    loop
  }
}

function mapDispatchToProps (dispatch) {
  return bindActionCreators({
    setLoop
  }, dispatch)
}

class Repeat extends React.Component {
  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick () {
    const { loop, setLoop } = this.props

    setLoop(!loop)
    saveState('loop', !loop)
  };

  render () {
    const { loop } = this.props

    return (
      <button onClick={this.handleClick}>
        { loop
          ? <i className='fa fa-repeat active' aria-hidden='true' />
          : <i className='fa fa-repeat' aria-hidden='true' />
        }
      </button>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Repeat)
