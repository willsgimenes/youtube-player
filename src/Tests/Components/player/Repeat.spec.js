/* global describe */
/* global beforeEach */
/* global it */
/* global expect */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import Repeat from '../../../components/player/Repeat'

function storageMock () {
  let storage = {}

  return {
    setItem: function (key, value) {
      storage[key] = value || ''
    },
    getItem: function (key) {
      return key in storage ? storage[key] : null
    },
    removeItem: function (key) {
      delete storage[key]
    },
    get length () {
      return Object.keys(storage).length
    },
    key: function (i) {
      let keys = Object.keys(storage)
      return keys[i] || null
    }
  }
}

window.localStorage = storageMock()

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

Enzyme.configure({ adapter: new Adapter() })

describe('Repeat button', () => {
  let wrapper, store, initialState

  beforeEach(() => {
    initialState = {
      player: {
        loop: true
      }
    }

    store = mockStore(initialState)
    wrapper = shallow(<Repeat store={store} />)
  })

  it('should render properly', () => {
    expect(wrapper.dive().find('button')).toHaveLength(1)
  })

  it('should have loop prop', () => {
    expect(wrapper.props().loop).toBeDefined()
  })

  it('should not have class if loop === false', () => {
    initialState = {
      player: {
        loop: false
      }
    }

    store = mockStore(initialState)
    wrapper = shallow(<Repeat store={store} />).dive()
    expect(wrapper.find('i').hasClass('fa fa-repeat active')).toBe(false)
    expect(wrapper).toMatchSnapshot()
  })

  it('should have class if loop === true', () => {
    initialState = {
      player: {
        loop: true
      }
    }

    store = mockStore(initialState)
    wrapper = shallow(<Repeat store={store} />).dive()
    expect(wrapper.find('i').hasClass('fa fa-repeat active')).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('should simulate click', () => {
    const mockAction = [
      {
        payload: false,
        type: 'SET_LOOP'
      }
    ]

    wrapper = shallow(<Repeat store={store} />)
    wrapper.dive().find('button').simulate('click')
    const actions = wrapper.props().store.getActions()
    expect(actions).toEqual(mockAction)
  })
})
