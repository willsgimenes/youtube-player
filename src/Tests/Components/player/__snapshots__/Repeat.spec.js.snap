// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repeat button should have class if loop === true 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Repeat
    loop={true}
    setLoop={[Function]}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        aria-hidden="true"
        className="fa fa-repeat active"
/>,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-hidden": "true",
        "className": "fa fa-repeat active",
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          aria-hidden="true"
          className="fa fa-repeat active"
/>,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-repeat active",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "storeSubscription": undefined,
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`Repeat button should not have class if loop === false 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Repeat
    loop={false}
    setLoop={[Function]}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    storeSubscription={
        Subscription {
            "listeners": Object {
              "clear": [Function],
              "get": [Function],
              "notify": [Function],
              "subscribe": [Function],
            },
            "onStateChange": [Function],
            "parentSub": undefined,
            "store": Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            },
            "unsubscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <i
        aria-hidden="true"
        className="fa fa-repeat"
/>,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-hidden": "true",
        "className": "fa fa-repeat",
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <i
          aria-hidden="true"
          className="fa fa-repeat"
/>,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-hidden": "true",
          "className": "fa fa-repeat",
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "storeSubscription": undefined,
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
